### PR TITLE
chore(ux): show registry error and hint for dockehub

### LINKED ISSUE #1216: improve error message when namespace is not specified in the reference of Dockerhub artifacts
Users might forget to add the namespace (e.g. `library` #1151) when pulling or copying Dockerhub images and get unexpected 401. Need to improve the error message and suggest user to add namespace.

Current behavior
```console
$ oras pull docker.io/nginx:latest
Error: failed to resolve latest: GET "https://registry-1.docker.io/v2/nginx/manifests/latest": response status code 401: unauthorized: authentication required: [map[Action:pull Class: Name:nginx Type:repository]]
```

Suggested error message:

```console
$ oras pull docker.io/nginx:latest
Error response from registry: pull access denied for docker.io/nginx:latest : response status code 401: unauthorized: requested access to the resource is denied
Namespace is missing, do you mean `oras pull docker.io/library/nginx:latest`? 
```

### PR DESCRIPTION
**Pull request by qweeah**
**What this PR does / why we need it**:
This PR 
1) adds prefix to errors returned by registry. E.g. invalid creds for fetching a manifest
```console
$ oras manifest fetch-config localhost:7000/repo:tag -u hello -p 123 >>
WARNING! Using --password via the CLI is insecure. Use --password-stdin.
Error response from registry: unauthorized: authentication required
```

2) gives recommendation if namespace is missing in a Dockerhub reference.
```console
$ oras pull docker.io/apline:latest
Error response from registry: failed to resolve latest: unauthorized: authentication required: [map[Action:pull Class: Name:apline Type:repository]]
Namespace is missing, do you mean `oras pull docker.io/library/apline:latest`?
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1216

**Please check the following list**:
- [x]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

### CODE REVIEW

File: cmd/oras/internal/errors/errors.go [hunk 1]
```diff
@@ -60,6 +65,51 @@ func CheckArgs(checker func(args []string) (bool, string), Usage string) cobra.P
...
+// Command returns an error-handled cobra command.
+func Command(cmd *cobra.Command, handler Handler) *cobra.Command {
+	runE := cmd.RunE
+	cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		err := runE(cmd, args)
+		if err == nil {
+			return nil
+		}
+		processor, err := handler.Handle(err, cmd)
+		if processor == nil {
```
Comment (shizhMSFT) [inline cmd/oras/internal/errors/errors.go:89]:
```suggestion
		if err != nil {
```

File: cmd/oras/internal/errors/errors.go [hunk 2]
```diff
@@ -60,6 +65,51 @@ func CheckArgs(checker func(args []string) (bool, string), Usage string) cobra.P
...
+			return err
+		}
+		return processor.Process(err, cmd.CommandPath())
```
Comment (shizhMSFT) [inline cmd/oras/internal/errors/errors.go:92]: The design of `Handler` and `Processor` seems can be simplified.

Here's the analysis:
1. The `err` is returned by `handler.Handle()`. That implies `Handle` has the information of `err`.
2. Similarly, `Handle` has the information of `processor`.

Combine 1 and 2, we can simplify the above code as
```go
processor := handler.Handle(err, cmd)
processor.Process(cmd.CommandPath())
```
or simply
```go
return handler.Handle(err, cmd).Process(cmd.CommandPath())
```
which can be further simplified as
```go
type ErrorHandler interface {
	HandleError(cmd *cobra.Command, err error) error
}
```
and
```go
return handler.HandleError(cmd, err)
```

File: cmd/oras/internal/option/remote.go [hunk 3]
```diff
@@ -322,3 +326,32 @@ func (opts *Remote) isPlainHttp(registry string) bool {
...
+}
+
+// Process processes error into oerrors.Error.
+func (opts *Remote) Process(err error, _ string) *oerrors.Error {
+	ret := oerrors.Error{
+		Err: err,
+	}
+	var errResp *errcode.ErrorResponse
+	if errors.As(err, &errResp) {
+		ret.Err = oerrors.GetInner(err, errResp)
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/remote.go:354]: This pattern seems can be further refactored.

File: cmd/oras/internal/errors/errors.go (hunk 1 above)
Comment (qweeah) [reply cmd/oras/internal/errors/errors.go]: refactored

File: cmd/oras/internal/errors/errors.go (hunk 2 above)
Comment (qweeah) [reply cmd/oras/internal/errors/errors.go]: Tried to use `Hande` and `Process` to decouple the logic of setting output prefix, scrubbing error info and give recommendation for know errors. I have merged the two interfaces.

File: cmd/oras/internal/option/remote.go (hunk 3 above)
Comment (qweeah) [reply cmd/oras/internal/option/remote.go]: Could you explain more detail on how to do that?

File: cmd/oras/internal/errors/errors.go [hunk 4]
```diff
@@ -60,6 +65,45 @@ func CheckArgs(checker func(args []string) (bool, string), Usage string) cobra.P
...
+	runE := cmd.RunE
+	cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		err := runE(cmd, args)
+		if err == nil {
+			return nil
+		}
+		if handled := handler.Handle(err, cmd); handled != nil {
+			return handled
+		}
+		return err
```
Comment (shizhMSFT) [inline cmd/oras/internal/errors/errors.go:86]:
```suggestion
		err := runE(cmd, args)
		if err != nil {
			return handler.Handle(err, cmd)
		}
		return nil
```

File: cmd/oras/internal/errors/errors.go [hunk 5]
```diff
@@ -60,6 +65,45 @@ func CheckArgs(checker func(args []string) (bool, string), Usage string) cobra.P
 	}
 }
 
+// Handler handles error during cmd execution.
+type Handler interface {
+	// Handle handles error during cmd execution.
+	// if nil is returned, then err cannot be handled.
+	Handle(err error, cmd *cobra.Command) error
```
Comment (shizhMSFT) [inline cmd/oras/internal/errors/errors.go:72]: nit: naming convention
```suggestion
	Handle(cmd *cobra.Command, err error) error
```

Comment (shizhMSFT) [reply cmd/oras/internal/errors/errors.go:72]: Example: https://pkg.go.dev/io/fs#WalkDirFunc
```go
type WalkDirFunc func(path string, d DirEntry, err error) error
```

File: cmd/oras/internal/errors/errors.go [hunk 6]
```diff
@@ -60,6 +65,45 @@ func CheckArgs(checker func(args []string) (bool, string), Usage string) cobra.P
...
+		return err
+	}
+	return cmd
+}
+
+// GetInner gets the inner error from the error response.
+func GetInner(err error, errResp *errcode.ErrorResponse) error {
+	inner := errResp.Errors
+	if len(inner) == 0 {
+		return fmt.Errorf("empty response body: %w", errResp)
```
Comment (shizhMSFT) [inline cmd/oras/internal/errors/errors.go:95]: `len(inner) == 0` does not imply empty response body but unrecognised response body.

See code: https://github.com/mostoja/oras-go/blob/9cac41930202c5f5fd8f5a2041862d593571c564/registry/remote/internal/errutil/errutil.go#L33-L48

File: cmd/oras/internal/errors/errors.go (hunk 4 above)
Comment (shizhMSFT) [reply cmd/oras/internal/errors/errors.go:86]: In fact, this handler is not a handler but a modifier.

Therefore, consider
```go
type Modifier interface {
    Modify(cmd *cobra.Command, err error) error
}

File: cmd/oras/internal/errors/errors.go [hunk 7]
```diff
@@ -60,6 +65,45 @@ func CheckArgs(checker func(args []string) (bool, string), Usage string) cobra.P
...
+	} else {
+		errContent := err.Error()
+		errRespContent := errResp.Error()
+		if idx := strings.Index(errContent, errRespContent); idx > 0 {
+			// remove HTTP related info
+			return fmt.Errorf("%s%w", errContent[:idx], inner)
+		}
```
Comment (shizhMSFT) [inline cmd/oras/internal/errors/errors.go:102]: Will it be safer to use `strings.Replace`?

File: cmd/oras/internal/option/remote.go [hunk 8]
```diff
@@ -322,3 +326,18 @@ func (opts *Remote) isPlainHttp(registry string) bool {
...
+		cmd.SetErrPrefix(oerrors.RegistryErrorPrefix)
+		return err
+	} else if errors.As(err, &errResp) {
+		cmd.SetErrPrefix(oerrors.RegistryErrorPrefix)
+		return &oerrors.Error{
+			Err: oerrors.GetInner(err, errResp),
+		}
+	}
+	return nil
+}
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/remote.go:343]: Followed by above comments, consider
```suggestion
// Modify modifies error during cmd execution.
func (opts *Remote) Modify(cmd *cobra.Command, err error) error {
	if errors.Is(err, errdef.ErrNotFound) {
		cmd.SetErrPrefix(oerrors.RegistryErrorPrefix)
		return err
	}

	var errResp *errcode.ErrorResponse
	if errors.As(err, &errResp) {
		cmd.SetErrPrefix(oerrors.RegistryErrorPrefix)
		return &oerrors.Error{
			Err: oerrors.GetInner(err, errResp),
		}
	}

	return err
}
```

File: cmd/oras/internal/errors/errors.go [hunk 9]
```diff
@@ -60,6 +65,45 @@ func CheckArgs(checker func(args []string) (bool, string), Usage string) cobra.P
...
+		if handled := handler.Handle(err, cmd); handled != nil {
+			return handled
+		}
+		return err
+	}
+	return cmd
+}
+
+// GetInner gets the inner error from the error response.
+func GetInner(err error, errResp *errcode.ErrorResponse) error {
```
Comment (shizhMSFT) [inline cmd/oras/internal/errors/errors.go:92]: Based on the code, this method is not to get the inner error from the error response but trim the error message. A proper name should be given to this method.

File: cmd/oras/internal/option/target.go [hunk 10]
```diff
@@ -269,3 +311,11 @@ func (opts *BinaryTarget) Parse() error {
 	opts.To.resolveFlag = append(opts.resolveFlag, opts.To.resolveFlag...)
 	return Parse(opts)
 }
+
+// Handle handles error during cmd execution.
+func (opts *BinaryTarget) Handle(err error, cmd *cobra.Command) error {
+	if err := opts.From.Handle(err, cmd); err != nil {
+		return err
+	}
+	return opts.To.Handle(err, cmd)
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/target.go:320]: Can we get line 320 covered?

File: cmd/oras/internal/errors/errors.go (hunk 4 above)
Comment (qweeah) [reply cmd/oras/internal/errors/errors.go:86]: > return handler.Handle(err, cmd)

This is wrong, if the function returns nil, then the original `err` must be returned.

File: cmd/oras/internal/errors/errors.go (hunk 6 above)
Comment (qweeah) [reply cmd/oras/internal/errors/errors.go:95]: What do you suggest as a proper output? 
How about `no error code found: `? 
I suggested to change the prefix to `Empty error response from registry` but rejected by @FeynmanZhou. He thinks it's too complicated for user to know three different kind of prefixes.

File: cmd/oras/internal/errors/errors.go (hunk 7 above)
Comment (qweeah) [reply cmd/oras/internal/errors/errors.go:102]: `strings.Replace` cannot exclude the case when `idx == 0`. Also it requires extra code to do error wrapping.

File: cmd/oras/internal/option/target.go (hunk 10 above)
Comment (qweeah) [reply cmd/oras/internal/option/target.go]: Marked. Will add test to make it covered once the refactoring is done.

File: cmd/oras/internal/errors/errors.go [hunk 11]
```diff
@@ -60,6 +65,60 @@ func CheckArgs(checker func(args []string) (bool, string), Usage string) cobra.P
...
+	return inner
+}
+
+// reWrap re-wraps errA to errC and trims out errB, returns nil if scrub fails.
+// +---------- errA ----------+
+// |         +---- errB ----+ |      +---- errA ----+
+// |         |    errC      | |  =>  |     errC     |
+// |         +--------------+ |      +--------------+
+// +--------------------------+
+func reWrap(errA error, errB error, errC error) error {
```
Comment (shizhMSFT) [inline cmd/oras/internal/errors/errors.go:111]: nit:
```suggestion
func reWrap(errA, errB, errC error) error {
```

File: cmd/oras/internal/option/target.go [hunk 12]
```diff
@@ -239,8 +244,54 @@ func (opts *Target) EnsureReferenceNotEmpty() error {
...
+		cmd.SetErrPrefix(oerrors.RegistryErrorPrefix)
+		ret := &oerrors.Error{
+			Err: oerrors.Trim(err, errResp),
+		}
+
+		if ref.Registry == "docker.io" && errResp.StatusCode == http.StatusUnauthorized {
+			if ref.Repository != "" && !strings.Contains(ref.Repository, "/") {
+				// docker.io/xxx -> docker.io/library/xxx
+				ref.Repository = "library/" + ref.Repository
+				ret.Recommendation = fmt.Sprintf("Namespace is missing, do you mean `%s %s`?", cmd.CommandPath(), ref)
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/target.go:284]: nit:
```suggestion
				ret.Recommendation = fmt.Sprintf("Namespace seems missing. Do you mean `%s %s`?", cmd.CommandPath(), ref)
```

**Review by shizhMSFT — APPROVED** [review]
LGTM with suggestions

**Pull request merged**